Create Squirrel

ArtemLazarchuk opened the pull request:
# Squirrel

<img width="1314" height="916" alt="image" src="https://github.com/user-attachments/assets/9e2f17bc-c9e2-4090-b07b-17c457125699" />

<img width="1313" height="909" alt="image" src="https://github.com/user-attachments/assets/44d32216-ddb7-4f8b-acb5-a7a42e406f9c" />

<img width="1313" height="914" alt="image" src="https://github.com/user-attachments/assets/f459389e-9b56-47cd-8888-6832fda80441" />

## One-Click Development Environment

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?hide_repo_select=true&ref=main&repo=Ivy-Interactive%2FIvy-Examples&machine=standardLinux32gb&devcontainer_path=.devcontainer%2Fsquirrel%2Fdevcontainer.json&location=EuropeWest)

Click the badge above to open this example in GitHub Codespaces with:
- **.NET 9.0** SDK pre-installed
- **Ready-to-run** environment (no local setup)

## Created Using Ivy

Web application created using [Ivy-Framework](https://github.com/Ivy-Interactive/Ivy-Framework).

**Ivy** unifies front-end and back-end into a single C# codebase for building internal tools and dashboards.

## What This Application Does

This example demonstrates data processing using the [Squirrel](https://github.com/sudipto80/Squirrel) library integrated with Ivy:

- **Load CSV**: Reads `fashion_products.csv` into a Squirrel `Table`
- **Filter & Sort**: Interactive controls for rating, price, brand, category, and sort order
- **Column Visibility**: Toggle which columns are shown/exported
- **Export Filtered CSV**: Download the current filtered view
- **Pagination**: Navigate results with page controls (default page size: 25)
- **Charts**: Brand-level line chart showing counts per brand and overall average for a selected product
- **Details Table**: Tabular breakdown of brand counts for the selected product

## Technical Implementation

- Uses Squirrel `Table` APIs for CSV load, sort, and iteration
- UI built with Ivy components and reactive state (`UseState`, `UseEffect`)
- Pagination via `Pagination` widget bound to state (top/bottom of the table)
- Chart built with Ivy `LineChart` using dimensions/measures
- Key files:
  - `squirrel/Apps/SquirrelCsvApp.cs` – data editor with filtering, sorting, pagination, and CSV export
  - `squirrel/Apps/PhysicsSimulationApp.cs` – chart and details view for brand counts

## How to Run

1. Prerequisites: **.NET 9.0 SDK**
2. Navigate to this example:
   ```bash
   cd squirrel
   ```
3. Restore dependencies:
   ```bash
   dotnet restore
   ```
4. Run the application:
   ```bash
   dotnet watch
   ```
5. Open your browser to the URL shown in the terminal (typically `http://localhost:5010`).

## How to Deploy

1. Navigate to this example:
   ```bash
   cd squirrel
   ```
2. Deploy to Ivy hosting:
   ```bash
   ivy deploy
   ```

## Learn More

- Squirrel library: `https://github.com/sudipto80/Squirrel`
- Ivy Documentation: `https://docs.ivy.app`

sudipto80 commented on squirrel/Apps/PhysicsSimulationApp.cs:
```diff
@@ -0,0 +1,156 @@
+        // Process data using Squirrel Table's row access
+        foreach (var row in originalTable.Rows)
+        {
+            var brand = Convert.ToString(row["Brand"]) ?? "";
```
This ToString and the next one are not needed because `row[X]` gives a string already

sudipto80 commented on squirrel/Apps/PhysicsSimulationApp.cs:
```diff
@@ -0,0 +1,156 @@
+        
+        // Create detailed data with brand breakdown for each product using Squirrel Table
+        // Use Squirrel Table's Rows collection to process data
+        var detailedData = new Dictionary<string, Dictionary<string, int>>();
```
This can be created using nested calls to `SplitOn` like this 

```csharp
  var brands = fashions.SplitOn("Brand");
        var detailedData = new Dictionary<string, Dictionary<string, int>>();
        brands.Select(t =>
                new
                {
                    Brand = t.Key,
                    Products = t.Value.SplitOn("Product Name")
                        .Select(z => new { Product = z.Key, Count = z.Value.RowCount })
                }
            ).ToList()
            .ForEach(z =>
            {
               detailedData.Add(z.Brand, new Dictionary<string, int>());
                foreach (var product in z.Products)
                {
                    detailedData[z.Brand].Add(product.Product, product.Count);
                }
            });
```

sudipto80 commented on squirrel/Apps/PhysicsSimulationApp.cs:
```diff
@@ -0,0 +1,156 @@
+            ? detailedData[selectedProduct.Value]
+            : new Dictionary<string, int>();
+        
+        var brandNames = selectedProductData.Keys.OrderBy(b => b).ToList();
```
To get the values of a column, just use the column name as the index on the `Squirrel.Table` instance like 

```csharp
var brandNames = tab["Brand"]
```

ArtemLazarchuk commented:
@sudipto80 Thank you for the feedback! I’ll fix it and let you know once it’s done.

ArtemLazarchuk commented:
@sudipto80 Could you please review it again?

sudipto80 reviewed (changes requested):
I have left couple of more comments around converting Squirrel table to strongly typed collection of products and saving to CSV. There is a old cheatsheet created for Squirrel. Might help 

Here is a cheatsheet. I created long time ago 
https://www.slideshare.net/slideshow/squirrel-do-morewithlesscodelightcheatsheet/48221385 

Here is the darker version
https://www.slideshare.net/slideshow/squirrel-do-morewithlesscodecheatsheet1/48185568

Also try this code with your fashion data (change the path) and use the latest release version. There will be a Nuget soon but the current Nuget is old 

```csharp

using Squirrel;

namespace IvyFashion;

class Program
{
    static void Main(string[] args)
    {
        var fashions = DataAcquisition.LoadCsv(
            @"/Users/sudiptamukherjee/Documents/GitHub/Squirrel/SquirrelSolution/IvyFashion/fashion.csv");
        
        // Brand and Product Counts

        var brands = fashions.SplitOn("Brand");
        
        // How many products are there for each brand
        brands.Select ( t =>
            {
                var prices = 
                    t.Value["Price"].Select(z => Convert.ToDecimal(z.Trim()))
                        .ToList();
                return new
                {
                    Brand = t.Key,
                    Count = t.Value.RowCount,
                    // Highest and lowest price item for this brand
                    HighestPrice = prices.Max(),
                    LowestPrice = prices.Min()
                };
            })
         //Create a new table with the new columns from the Select projection 
        .ToTableFromAnonList()
        .PrettyDump(header: "Brand and Product Counts", rowColor: ConsoleColor.Blue);
        
        
        // Filter by product then by size 
        fashions.Filter("Product Name", "Dress")
                .SortInThisOrder("Size", ["S","M","L","XL"])
                .PrettyDump(header: "Dress", rowColor: ConsoleColor.Magenta);
        
        // Filter on multiple columns 
        fashions.Filter("Brand", "Adidas")
                .Filter("Product Name", "Dress")
                .Filter("Size", "M")
                .PrettyDump(header: "Adidas Dress M", rowColor: ConsoleColor.Magenta);
        
        // You can also filter by regex. find anything that has the word "Women" in it
        fashions.FilterByRegex("Category", "Women")
            .SortBy("Rating", how: SortDirection.Descending)
            .PrettyDump(header: "Women's fashion", rowColor: ConsoleColor.Red);;
        
        // Add column programmatically 
        fashions.AddColumn(columnName: "ValueForMoney", formula: "[Rating]/[Price]", decimalDigits: 2);
        var valueForMoney = fashions.SortBy("ValueForMoney", how: SortDirection.Descending)
            // Pick the columns we want 
            .Pick("Product Name","Brand","Category","Price", "Rating", "ValueForMoney");
            // Dump to console 
            valueForMoney
                //Pick the top 15
                .Top(15)
                .PrettyDump(header: "Top 5 fashion for money", rowColor: ConsoleColor.Green);
            
            valueForMoney
                //Pick the bottom 15
                .Bottom(15)
                .PrettyDump(header: "Bottom 5 fashion for money", rowColor: ConsoleColor.DarkRed);
    }
}
```

ArtemLazarchuk commented:
@sudipto80 Thanks for the hint! I’ll check it out and make use of it.

ArtemLazarchuk commented:
@sudipto80 could you please take another look?

I’ve addressed your comments, fixed those issues, and also added a demo application for the code you provided.

sudipto80 reviewed (approved):
Looks great! Thanks!
